src/os/Process.php

```diff
@@ -0,0 +1,122 @@
+      const type TProcessID = _OS\PID;
+      const type TExitCode = _OS\ExitCode;
+      private function __construct(
+        private self::TProcessID $pid,
+        private IO\CloseableReadFDHandle $closeNotifier,
+      ) {}
+
+      public function getPID(): self::TProcessID {
+        return $this->pid;
+      }
+
+      <<__Memoize>>
+      public async function getExitCodeAsync(): Awaitable<self::TExitCode> {
+        await $this->closeNotifier->readAllAsync();
+        $status = null;
+        \pcntl_waitpid($this->pid, inout $status);
+        return \pcntl_wexitstatus($status);
+      }
+
+      public function close(): void {
+        $status = null;
+        switch (
+          \pcntl_waitpid($this->pid, inout $status, \WNOHANG | \WUNTRACED)
+        ) {
+          case 0:
+            // The process has not been exited. Kill it.
+            \posix_kill($this->pid, \SIGKILL);
+            \pcntl_waitpid($this->pid, inout $status, \WUNTRACED);
+            break;
+          default:
+            // No clean up needed because the process has been exited
+            break;
+        }
+        $this->closeNotifier->close();
+      }
+
+      <<__ReturnDisposable>>
+      final public function closeWhenDisposed(): \IDisposable {
+        return new \HH\Lib\_Private\_IO\CloseWhenDisposed($this);
+      }
+
+      public static function forkAndExec(
+        vec<string> $argv,
+        dict<string, string> $environment_variable = dict[],
+        ?IO\ReadFDHandle $stdin = null,
+        ?IO\WriteFDHandle $stdout = null,
+        ?IO\WriteFDHandle $stderr = null,
+        this::TForkAndExecOptions $options = shape(),
+        string $path = $argv[0],
+      ): Process {
+        list($read, $write) = IO\pipe();
+        using $write->closeWhenDisposed();
+        $fds = dict[
+          \HH\Lib\_Private\_OS\STDIN_FILENO =>
+            $stdin?->getFileDescriptor() ?? OS\stdin(),
+          \HH\Lib\_Private\_OS\STDOUT_FILENO =>
+            $stdout?->getFileDescriptor() ?? OS\stdout(),
+          \HH\Lib\_Private\_OS\STDERR_FILENO =>
+            $stderr?->getFileDescriptor() ?? OS\stderr(),
+          _OS\CLOSE_NOTIFIER_FILENO => $write->getFileDescriptor(),
+        ];
+
+        return new Process(
+          \HH\Lib\_Private\_OS\fork_and_execve(
+            $path,
+            $argv,
+            Vec\map_with_key($environment_variable, ($k, $v) ==> {
+              // In the shell command language, a word consisting solely of
+              // underscores, digits, and alphabetics from the portable character
+              // set. The first character of a name is not a digit.
+              // See https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_235
+              invariant(
+                Regex\matches($k, re"/[a-zA-Z_][a-zA-Z0-9_]*/"),
+                'Illegal environment variable name: %s',
+                $k,
+              );
+              return $k.'='.$v;
```

Comment:
The `$v` in the scope is not sanitized or processed. Is the fork_and_execve API safe when passed mal(formed|licious) environment variable values?

Reply:
Good question. `$v` should be able to contain almost any characters, including `=`. I think the only character that could have a special meaning in `$v` is `\0`, which makes the string truncated. I can add a test to confirm if my understanding is correct.